### PR TITLE
fix: Fix delete orphan draft pages when the space is deleted - EXO-70385 - Meeds-io/MIPs#119

### PR DESCRIPTION
**Pull request by hakermi**
Prior to this change, when deleting a note space, the related space note data is not well deleted when the space have orphan drafts (for non existing pages) which raises an exception related to the constraint foreign key violation when attempting to delete note pages with children orphan drafts.
This PR makes sure to delete all note drafts before proceeding to delete the related note pages.

### CODE REVIEW

File: notes-service/src/main/java/org/exoplatform/wiki/jpa/dao/DraftPageDAO.java [hunk 1]
```diff
@@ -89,6 +90,13 @@ public DraftPageEntity findLatestDraftPageByTargetPage(Long targetPageId) {
     }
   }
 
+  @ExoTransactional
+  public void deleteNoteOrphanDraftPages(long parentPageId) {
```
Comment (azayati) [inline notes-service/src/main/java/org/exoplatform/wiki/jpa/dao/DraftPageDAO.java]:
```suggestion
  public void deleteOrphanDraftPagesByParentPage(long parentPageId) {
```

File: notes-service/src/main/java/org/exoplatform/wiki/jpa/dao/DraftPageDAO.java [hunk 2]
```diff
@@ -89,6 +90,13 @@ public DraftPageEntity findLatestDraftPageByTargetPage(Long targetPageId) {
...
+    Query query = getEntityManager().createNamedQuery("wikiDraftPage.deleteNoteOrphanDraftPages");
```
Comment (azayati) [inline notes-service/src/main/java/org/exoplatform/wiki/jpa/dao/DraftPageDAO.java]:
```suggestion
    Query query = getEntityManager().createNamedQuery("wikiDraftPage.deleteOrphanDraftPagesByParentPage");
```

File: notes-service/src/main/java/org/exoplatform/wiki/jpa/entity/DraftPageEntity.java [hunk 3]
```diff
@@ -53,7 +53,8 @@
         @NamedQuery(name = "wikiDraftPage.findDraftPageByTargetPage", query = "SELECT d FROM WikiDraftPageEntity d WHERE d.targetPage.id = :targetPageId"),
         @NamedQuery(name = "wikiDraftPage.findDraftPagesByParentPage", query = "SELECT d FROM WikiDraftPageEntity d WHERE d.parentPage.id = :parentPageId"),
         @NamedQuery(name = "wikiDraftPage.findLatestDraftPageByTargetPageAndLang", query = "SELECT d FROM WikiDraftPageEntity d WHERE d.targetPage.id = :targetPageId AND " +
-                                                                                                  "((:lang IS NULL AND d.lang IS NULL) OR (:lang IS NOT NULL AND d.lang = :lang)) ORDER BY d.updatedDate DESC"), })
+                                                                                                  "((:lang IS NULL AND d.lang IS NULL) OR (:lang IS NOT NULL AND d.lang = :lang)) ORDER BY d.updatedDate DESC"),
+        @NamedQuery(name = "wikiDraftPage.deleteNoteOrphanDraftPages", query = "DELETE FROM WikiDraftPageEntity d WHERE d.targetPage IS NULL AND d.parentPage.id=:id")})
```
Comment (azayati) [inline notes-service/src/main/java/org/exoplatform/wiki/jpa/entity/DraftPageEntity.java]:
```suggestion
        @NamedQuery(name = "wikiDraftPage.deleteOrphanDraftPagesByParentPage", query = "DELETE FROM WikiDraftPageEntity d WHERE d.targetPage IS NULL AND d.parentPage.id=:id")})
```

File: notes-service/src/main/java/org/exoplatform/wiki/jpa/JPADataStorage.java [hunk 4]
```diff
@@ -1559,4 +1559,12 @@ public void deleteVersionsByNoteIdAndLang(Long noteId, String lang) throws WikiE
     pageEntity.setVersions(history);
     pageDAO.update(pageEntity);
   }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void deleteNoteOrphanDraftPages(long parentPageId) {
```
Comment (azayati) [inline notes-service/src/main/java/org/exoplatform/wiki/jpa/JPADataStorage.java]:
```suggestion
  public void deleteOrphanDraftPagesByParentPage(long parentPageId) {
```

File: notes-service/src/main/java/org/exoplatform/wiki/jpa/JPADataStorage.java [hunk 5]
```diff
@@ -1559,4 +1559,12 @@ public void deleteVersionsByNoteIdAndLang(Long noteId, String lang) throws WikiE
...
+    draftPageDAO.deleteNoteOrphanDraftPages(parentPageId);
```
Comment (azayati) [inline notes-service/src/main/java/org/exoplatform/wiki/jpa/JPADataStorage.java]:
```suggestion
    draftPageDAO.deleteOrphanDraftPagesByParentPage(parentPageId);
```

File: notes-service/src/main/java/org/exoplatform/wiki/service/NoteService.java [hunk 6]
```diff
@@ -755,4 +755,13 @@ Page getNoteByIdAndLang(Long pageId, Identity userIdentity, String source, Strin
...
   void deleteVersionsByNoteIdAndLang(Long noteId, String username, String lang) throws WikiException;
+
+
+  /**
+   * Delete all drafts those are not assigned to created page
+   * instead they are assigned to the Note home page
+   *
+   * @param parentPageId Note parent page id
+   */
+  void deleteNoteOrphanDraftPages(long parentPageId);
```
Comment (azayati) [inline notes-service/src/main/java/org/exoplatform/wiki/service/NoteService.java]:
```suggestion
  void removeOrphanDraftPagesByParentPage(long parentPageId);
```

File: notes-service/src/main/java/org/exoplatform/wiki/service/NoteService.java [hunk 7]
```diff
@@ -755,4 +755,13 @@ Page getNoteByIdAndLang(Long pageId, Identity userIdentity, String source, Strin
   @Deprecated
   // Use {@link deleteVersionsByNoteIdAndLang(Long noteId, String lang)} instead
   void deleteVersionsByNoteIdAndLang(Long noteId, String username, String lang) throws WikiException;
+
+
+  /**
+   * Delete all drafts those are not assigned to created page
```
Comment (azayati) [inline notes-service/src/main/java/org/exoplatform/wiki/service/NoteService.java]:
```suggestion
   * Remove all children drafts of a parent page without existing target page
```

File: notes-service/src/main/java/org/exoplatform/wiki/service/NoteService.java [hunk 8]
```diff
@@ -755,4 +755,13 @@ Page getNoteByIdAndLang(Long pageId, Identity userIdentity, String source, Strin
...
+   * instead they are assigned to the Note home page
```
Comment (azayati) [inline notes-service/src/main/java/org/exoplatform/wiki/service/NoteService.java]: what means ?

File: notes-service/src/main/java/org/exoplatform/wiki/service/impl/NoteServiceImpl.java [hunk 9]
```diff
@@ -1754,4 +1754,12 @@ private Map<String, List<MetadataItem>> retrieveMetadataItems(String noteId, Str
                  });
     return metadata;
   }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void deleteNoteOrphanDraftPages(long parentPageId) {
```
Comment (azayati) [inline notes-service/src/main/java/org/exoplatform/wiki/service/impl/NoteServiceImpl.java]:
```suggestion
  public void removeOrphanDraftPagesByParentPage(long parentPageId) {
```

File: notes-service/src/main/java/org/exoplatform/wiki/service/DataStorage.java [hunk 10]
```diff
@@ -264,4 +264,11 @@ public default List<Attachment> getAttachmentsOfPage(Page page, boolean loadCont
    */
   void deleteVersionsByNoteIdAndLang(Long noteId, String lang) throws WikiException;
 
+  /**
+   * Delete all drafts those are not assigned to created page
+   * instead they are assigned to the Note home page
+   *
+   * @param parentPageId Note parent page id
+   */
+  void deleteNoteOrphanDraftPages(long parentPageId);
```
Comment (azayati) [inline notes-service/src/main/java/org/exoplatform/wiki/service/DataStorage.java]:
```suggestion
  void deleteOrphanDraftPagesByParentPage(long parentPageId);
```

File: notes-service/src/main/java/org/exoplatform/wiki/service/DataStorage.java [hunk 11]
```diff
@@ -264,4 +264,11 @@ public default List<Attachment> getAttachmentsOfPage(Page page, boolean loadCont
    */
   void deleteVersionsByNoteIdAndLang(Long noteId, String lang) throws WikiException;
 
+  /**
+   * Delete all drafts those are not assigned to created page
+   * instead they are assigned to the Note home page
+   *
+   * @param parentPageId Note parent page id
+   */
```
Comment (azayati) [inline notes-service/src/main/java/org/exoplatform/wiki/service/DataStorage.java]: put the same doc as Noteservice method

File: notes-service/src/test/java/org/exoplatform/wiki/jpa/JPADataStorageTest.java [hunk 12]
```diff
@@ -1142,6 +1145,26 @@ public void testGetWatchers() throws WikiException {
     assertTrue(step3Watchers.contains("user2"));
   }
 
+  public void testDeleteNoteOrphanDraftPages() throws Exception {
```
Comment (azayati) [inline notes-service/src/test/java/org/exoplatform/wiki/jpa/JPADataStorageTest.java]:
```suggestion
  public void testDeleteOrphanDraftPagesByParentPage() throws Exception {
```

File: notes-service/src/test/java/org/exoplatform/wiki/service/TestNoteService.java [hunk 13]
```diff
@@ -684,4 +685,18 @@ public void testGetPublishedVersionByPageIdAndLang() throws WikiException, Illeg
     noteService.createVersionOfNote(note, user.getUserId());
     assertNotNull(noteService.getPublishedVersionByPageIdAndLang(Long.valueOf(note.getId()), "en"));
   }
+
+  public void testDeleteNoteOrphanDraftPages() throws Exception {
```
Comment (azayati) [inline notes-service/src/test/java/org/exoplatform/wiki/service/TestNoteService.java]:
```suggestion
  public void testRemoveOrphanDraftPagesByParentPage() throws Exception {
```

File: notes-service/src/test/java/org/exoplatform/wiki/jpa/dao/DraftPageDAOTest.java [hunk 14]
```diff
@@ -346,6 +348,7 @@ public void testFindLatestDraftPageByTargetPage() {
     dp1.setCreatedDate(new Date());
     dp1.setUpdatedDate(new Date());
     draftPageDAO.create(dp1);
+    persist();
```
Comment (azayati) [inline notes-service/src/test/java/org/exoplatform/wiki/jpa/dao/DraftPageDAOTest.java:351]: this is necessary ?

**Review by azayati — CHANGES_REQUESTED** [review]

File: notes-service/src/test/java/org/exoplatform/wiki/jpa/dao/DraftPageDAOTest.java (hunk 14 above)
Comment (hakermi) [reply notes-service/src/test/java/org/exoplatform/wiki/jpa/dao/DraftPageDAOTest.java:351]: yes, that fixes a random fail in that test

**Review by azayati — APPROVED** [review]
LGTM

**Pull request merged**